File: app/rotas/laudos.py
"""/laudos - laudo tecnico. Sem data de validade (IN 15/2022, art. 10, SS3)."""

from __future__ import annotations

from datetime import date

from fastapi import APIRouter, Form, Request
from fastapi.responses import RedirectResponse
from sqlalchemy import select

from app.dependencias import SessaoDep, UsuarioDep
from app.modelos import (
    LaudoPosto,
    LaudoTecnico,
    ParecerTecnico,
    PostoTrabalho,
    ProfissionalHabilitado,
    TipoAdicional,
    UnidadeUorg,
)
from app.servicos import datas_br, parecer as servico_parecer
from app.servicos.nup import PADRAO as _PADRAO_NUP  # noqa: F401  (mantem simetria de import)
from app.web import numero_da_pagina, pagina, recortar

rotas = APIRouter(tags=["laudos"])

RODAPE_FILA = (
    "o laudo não tem prazo de validade (IN 15/2022, art. 10, §3º); esta lista é "
    "apenas fila de trabalho, não vencimento."
)


SITUACOES_LAUDO = (("VIGENTE", "Vigentes"), ("SUPERADO", "Superados"))


def _tela_lista(
    request: Request,
    s,
    usuario,
    situacao: str = "",
    q: str = "",
    *,
    erro: str | None = None,
    digitado: dict | None = None,
):
    """A fila, com o popup de cadastro em branco ou com o que foi digitado.

    Espelha `processos._tela_novo`: `digitado` nao entra na URL, e por isso a
    recusa renderiza a propria tela em vez de redirecionar. Antes ela trocava a
    fila inteira pela pagina de erro do sistema, e o "voltar" do navegador nao
    devolve formulario enviado por POST — tipo, unidade, subscritor, data e a
    marca de coletivo eram redigitados por causa da pontuacao de um numero.
    """
    todos = list(
        s.execute(select(LaudoTecnico).order_by(LaudoTecnico.numero_siape.desc())).scalars()
    )
    resumo = {codigo: 0 for codigo, _ in SITUACOES_LAUDO}
    for laudo in todos:
        resumo[laudo.status] = resumo.get(laudo.status, 0) + 1

    busca = q.strip().lower()
    laudos = [
        laudo
        for laudo in todos
        if (not situacao or laudo.status == situacao)
        and (
            not busca
            or busca in laudo.numero_siape.lower()
            or (laudo.unidade is not None and busca in laudo.unidade.nome_extenso.lower())
            or (laudo.subscritor is not None and busca in laudo.subscritor.nome.lower())
        )
    ]
    # O recorte entra ANTES do laco de `derivados`, e nao depois: aquele laco e
    # uma consulta por laudo — a unica consulta por linha que sobrou nas listas
    # deste sistema —, e com 200 laudos na fila eram 200 idas ao banco para
    # desenhar uma tela de 50 linhas. Paginando primeiro, sao 50. Filtrar antes
    # de recortar continua obrigatorio: a busca casa por unidade e subscritor,
    # que sao relacionamentos, e nao cabe na consulta de cima.
    recorte = recortar(laudos, numero_da_pagina(request))
    derivados = {}
    for laudo in recorte.itens:
        derivados[laudo.id] = s.execute(
            select(ParecerTecnico).where(ParecerTecnico.laudo_id == laudo.id)
        ).scalars().all()
    return pagina(
        request,
        "paginas/laudos.html",
        usuario=usuario,
        laudos=recorte.itens,
        recorte=recorte,
        derivados=derivados,
        situacoes=SITUACOES_LAUDO,
        filtro_situacao=situacao,
        resumo=resumo,
        total=len(todos),
        q=q,
        unidades=list(s.execute(select(UnidadeUorg).order_by(UnidadeUorg.nome_extenso)).scalars()),
        tipos=list(s.execute(select(TipoAdicional)).scalars()),
        subscritores=list(s.execute(select(ProfissionalHabilitado)).scalars()),
        rodape_fila=RODAPE_FILA,
        erro=erro,
        digitado=digitado or {},
    )


@rotas.get("/laudos")
def listar(
    request: Request,
    s: SessaoDep,
    usuario: UsuarioDep,
    situacao: str = "",
    q: str = "",
):
    """A fila dos laudos.

    Ate aqui era a unica lista longa do sistema SEM recorte nenhum: nem as
    fichas de situacao que /adicionais, /certificados, /turmas e a fila de EPI
    usam, nem a barra de filtros de /processos. Quem chegava com um numero na
    mao lia a lista inteira. O recorte e de UMA dimensao (a situacao do laudo,
    que a norma fecha em duas), entao ele e fileira de fichas com contagem, e o
    numero e a busca livre entram no `filtro-inline` ao lado — que e a regra
    escrita em `csso.css`, na fileira de fichas.

    A contagem sai da lista COMPLETA, e nao da filtrada: uma ficha "Superados
    (0)" que so diz zero porque o filtro vigente ja excluiu os superados nao
    informa nada e ainda contradiz a tela seguinte.
    """
    usuario.exigir("laudo.ver")
    return _tela_lista(request, s, usuario, situacao, q)


@rotas.post("/laudos")
def criar(
    request: Request,
    s: SessaoDep,
    usuario: UsuarioDep,
    numero_siape: str = Form(...),
    tipo_adicional_id: int = Form(...),
    unidade_uorg_id: int = Form(...),
    subscritor_id: str = Form(""),
    data_emissao: str = Form(""),
    coletivo: str = Form(""),
):
    usuario.exigir("laudo.criar")
    import re

    # Tudo o que foi digitado, guardado antes da primeira recusa.
    digitado = {
        "numero_siape": numero_siape,
        "tipo_adicional_id": tipo_adicional_id,
        "unidade_uorg_id": unidade_uorg_id,
        "subscritor_id": subscritor_id,
        "data_emissao": data_emissao,
        "coletivo": coletivo,
    }
    if not re.fullmatch(r"\d{5}-\d{3}\.\d{3}/\d{4}", numero_siape.strip()):
        return _tela_lista(
            request,
            s,
            usuario,
            erro="Número de laudo inválido — o formato do laudo SIAPE é 26255-000.125/2019.",
            digitado=digitado,
        )
    ano = int(numero_siape.strip()[-4:])
    laudo = LaudoTecnico(
        numero_siape=numero_siape.strip(),
        ano=ano,
        tipo_adicional_id=tipo_adicional_id,
        unidade_uorg_id=unidade_uorg_id,
        subscritor_id=int(subscritor_id) if subscritor_id else None,
        data_emissao=date.fromisoformat(data_emissao) if data_emissao else None,
        coletivo=coletivo == "1",
    )
    s.add(laudo)
    s.commit()
    return RedirectResponse(f"/laudos/{laudo.id}", status_code=303)


@rotas.get("/laudos/{laudo_id}")
def ficha(
    request: Request,
    s: SessaoDep,
    usuario: UsuarioDep,
    laudo_id: int,
    mensagem: str | None = None,
    erro: str | None = None,
):
    usuario.exigir("laudo.ver")
    laudo = s.get(LaudoTecnico, laudo_id)
    if laudo is None:
        return RedirectResponse("/laudos", status_code=303)
    derivados = list(
        s.execute(select(ParecerTecnico).where(ParecerTecnico.laudo_id == laudo.id)).scalars()
    )
    meses = (
        datas_br.meses_entre(laudo.data_ultima_conferencia, date.today())
        if laudo.data_ultima_conferencia
        else None
    )
    return pagina(
        request,
        "paginas/laudo_ficha.html",
        usuario=usuario,
        laudo=laudo,
        derivados=derivados,
        meses_desde_conferencia=meses,
        postos=list(s.execute(select(PostoTrabalho).order_by(PostoTrabalho.nome)).scalars()),
        rodape_fila=RODAPE_FILA,
        mensagem=mensagem,
        erro=erro,
    )


@rotas.post("/laudos/{laudo_id}/postos")
def vincular_posto(
    s: SessaoDep, usuario: UsuarioDep, laudo_id: int, posto_trabalho_id: int = Form(...)
):
    usuario.exigir("laudo.criar")
    ja = s.get(LaudoPosto, (laudo_id, posto_trabalho_id))
    if ja is None:
        s.add(LaudoPosto(laudo_id=laudo_id, posto_trabalho_id=posto_trabalho_id))
        s.commit()
    return RedirectResponse(f"/laudos/{laudo_id}", status_code=303)


@rotas.post("/laudos/{laudo_id}/conferencia")
def registrar_conferencia(
    s: SessaoDep, usuario: UsuarioDep, laudo_id: int, motivo: str = Form(...)
):
    usuario.exigir("laudo.criar")
    laudo = s.get(LaudoTecnico, laudo_id)
    if laudo is not None:
        laudo.data_ultima_conferencia = date.today()
        laudo.motivo_ultima_conferencia = motivo
        s.commit()
    return RedirectResponse(f"/laudos/{laudo_id}?mensagem=Conferência registrada.", status_code=303)


@rotas.post("/laudos/{laudo_id}/superar")
def superar(
    s: SessaoDep,
    usuario: UsuarioDep,
    laudo_id: int,
    motivo: str = Form(...),
    substituto_id: str = Form(""),
):
    laudo = s.get(LaudoTecnico, laudo_id)
    if laudo is None:
        return RedirectResponse("/laudos", status_code=303)
    substituto = s.get(LaudoTecnico, int(substituto_id)) if substituto_id else None
    derivados = servico_parecer.marcar_laudo_superado(s, laudo, usuario, motivo, substituto)
    s.commit()
    return RedirectResponse(
        f"/laudos/{laudo_id}?mensagem=Laudo superado; "
        f"{len(derivados)} parecer(es) em reavaliação.",
        status_code=303,
    )

File: nuvem/src/rotas/laudos.ts
/**
 * /laudos — laudo técnico. Sem data de validade (IN 15/2022, art. 10, §3º).
 * Porte de `app/rotas/laudos.py`.
 */
import { Hono } from "hono";
import { and, asc, desc, eq } from "drizzle-orm";
import type { Ambiente, Ctx } from "../nucleo/contexto.js";
import { ErroHttp } from "../nucleo/erros.js";
import { formulario, usuarioCom, usuarioLogado } from "../dependencias.js";
import { numeroDaPagina, pagina, recortar, redirecionar } from "../web.js";
import { dicionario } from "../dicionario.js";
import {
  laudo_posto,
  laudo_tecnico,
  parecer_tecnico,
  posto_trabalho,
  profissional_habilitado,
  tipo_adicional,
  unidade_uorg,
} from "../db/esquema/index.js";
import * as datas_br from "../servicos/datas_br.js";
import * as servico_parecer from "../servicos/parecer.js";
import type { UsuarioAtual } from "../servicos/rbac.js";

export const rotas = new Hono<Ambiente>();

export const RODAPE_FILA =
  "o laudo não tem prazo de validade (IN 15/2022, art. 10, §3º); esta lista é " +
  "apenas fila de trabalho, não vencimento.";

export const SITUACOES_LAUDO: readonly (readonly [string, string])[] = [
  ["VIGENTE", "Vigentes"],
  ["SUPERADO", "Superados"],
];

function _id(c: Ctx): number {
  const v = c.req.param("laudo_id");
  if (!v || !/^\d+$/.test(v)) throw new ErroHttp(422, "Parâmetro inválido: laudo_id");
  return Number(v);
}

/**
 * A fila, com o popup de cadastro em branco ou com o que foi digitado.
 * `digitado` não entra na URL, e por isso a recusa renderiza a própria tela em
 * vez de redirecionar.
 */
async function _tela_lista(
  c: Ctx,
  usuario: UsuarioAtual,
  situacao = "",
  q = "",
  o: { erro?: string | null; digitado?: Record<string, unknown> } = {},
) {
  const tx = c.get("tx");
  const todos = await tx.query.laudo_tecnico.findMany({
    with: { unidade: true, subscritor: true, tipo_adicional: true },
    orderBy: [desc(laudo_tecnico.numero_siape)],
  });
  const resumo: Record<string, number> = {};
  for (const [codigo] of SITUACOES_LAUDO) resumo[codigo] = 0;
  for (const laudo of todos) resumo[laudo.status] = (resumo[laudo.status] ?? 0) + 1;

  const busca = q.trim().toLowerCase();
  const laudos = todos.filter(
    (laudo) =>
      (!situacao || laudo.status === situacao) &&
      (!busca ||
        laudo.numero_siape.toLowerCase().includes(busca) ||
        (laudo.unidade !== null && laudo.unidade.nome_extenso.toLowerCase().includes(busca)) ||
        (laudo.subscritor !== null && laudo.subscritor.nome.toLowerCase().includes(busca))),
  );
  // O recorte entra ANTES do laço de `derivados`: aquele laço é uma consulta
  // por laudo, e paginando primeiro são 50, não 200.
  const recorte = recortar(laudos, numeroDaPagina(c));
  const derivados: Record<number, unknown[]> = {};
  for (const laudo of recorte.itens) {
    derivados[laudo.id] = await tx.select().from(parecer_tecnico).where(eq(parecer_tecnico.laudo_id, laudo.id));
  }
  return pagina(c, "paginas/laudos.html", usuario, {
    laudos: recorte.itens,
    recorte,
    derivados,
    situacoes: SITUACOES_LAUDO,
    filtro_situacao: situacao,
    resumo: dicionario(resumo),
    total: todos.length,
    q,
    unidades: await tx.select().from(unidade_uorg).orderBy(asc(unidade_uorg.nome_extenso)),
    tipos: await tx.select().from(tipo_adicional).orderBy(asc(tipo_adicional.id)),
    subscritores: await tx.select().from(profissional_habilitado).orderBy(asc(profissional_habilitado.id)),
    rodape_fila: RODAPE_FILA,
    erro: o.erro ?? null,
    digitado: dicionario(o.digitado ?? {}),
  });
}

/**
 * A fila dos laudos. A contagem das fichas sai da lista COMPLETA, e não da
 * filtrada: "Superados (0)" só porque o filtro excluiu os superados não
 * informaria nada.
 */
rotas.get("/laudos", async (c) => {
  const usuario = await usuarioCom(c, "laudo.ver");
  return _tela_lista(c, usuario, c.req.query("situacao") ?? "", c.req.query("q") ?? "");
});

rotas.post("/laudos", async (c) => {
  const usuario = await usuarioCom(c, "laudo.criar");
  const tx = c.get("tx");
  const f = await formulario(c);
  const tipo_adicional_id = f.inteiro("tipo_adicional_id");
  const unidade_uorg_id = f.inteiro("unidade_uorg_id");
  if (!f.tem("numero_siape") || tipo_adicional_id === null || unidade_uorg_id === null) {
    throw new ErroHttp(422, "Campo obrigatório ausente: numero_siape, tipo_adicional_id, unidade_uorg_id");
  }
  const numero_siape = f.texto("numero_siape");
  const subscritor_id = f.texto("subscritor_id");
  const data_emissao = f.texto("data_emissao");
  const coletivo = f.texto("coletivo");
  // Tudo o que foi digitado, guardado antes da primeira recusa.
  const digitado = { numero_siape, tipo_adicional_id, unidade_uorg_id, subscritor_id, data_emissao, coletivo };
  if (!/^\d{5}-\d{3}\.\d{3}\/\d{4}$/.test(numero_siape.trim())) {
    return _tela_lista(c, usuario, "", "", {
      erro: "Número de laudo inválido — o formato do laudo SIAPE é 26255-000.125/2019.",
      digitado,
    });
  }
  const [laudo] = await tx
    .insert(laudo_tecnico)
    .values({
      numero_siape: numero_siape.trim(),
      ano: Number(numero_siape.trim().slice(-4)),
      tipo_adicional_id,
      unidade_uorg_id,
      subscritor_id: subscritor_id ? Number(subscritor_id) : null,
      data_emissao: data_emissao || null,
      coletivo: coletivo === "1",
    })
    .returning();
  return redirecionar(c, `/laudos/${laudo!.id}`);
});

rotas.get("/laudos/:laudo_id{[0-9]+}", async (c) => {
  const usuario = await usuarioCom(c, "laudo.ver");
  const tx = c.get("tx");
  const laudo = await tx.query.laudo_tecnico.findFirst({
    where: eq(laudo_tecnico.id, _id(c)),
    with: {
      unidade: true,
      tipo_adicional: true,
      subscritor: true,
      postos: { with: { posto: { with: { unidade: true } } } },
    },
  });
  if (!laudo) return redirecionar(c, "/laudos");
  const derivados = await tx.query.parecer_tecnico.findMany({
    where: eq(parecer_tecnico.laudo_id, laudo.id),
    with: { servidor: true },
    orderBy: [asc(parecer_tecnico.id)],
  });
  const meses = laudo.data_ultima_conferencia
    ? datas_br.meses_entre(laudo.data_ultima_conferencia, datas_br.hoje())
    : null;
  return pagina(c, "paginas/laudo_ficha.html", usuario, {
    laudo,
    derivados,
    meses_desde_conferencia: meses,
    postos: await tx.query.posto_trabalho.findMany({ with: { unidade: true }, orderBy: [asc(posto_trabalho.nome)] }),
    rodape_fila: RODAPE_FILA,
    mensagem: c.req.query("mensagem") ?? null,
    erro: c.req.query("erro") ?? null,
  });
});

rotas.post("/laudos/:laudo_id{[0-9]+}/postos", async (c) => {
  await usuarioCom(c, "laudo.criar");
  const tx = c.get("tx");
  const laudo_id = _id(c);
  const posto_trabalho_id = (await formulario(c)).inteiro("posto_trabalho_id");
  if (posto_trabalho_id === null) throw new ErroHttp(422, "Campo obrigatório ausente: posto_trabalho_id");
  const [ja] = await tx
    .select()
    .from(laudo_posto)
    .where(and(eq(laudo_posto.laudo_id, laudo_id), eq(laudo_posto.posto_trabalho_id, posto_trabalho_id)));
  if (!ja) await tx.insert(laudo_posto).values({ laudo_id, posto_trabalho_id });
  return redirecionar(c, `/laudos/${laudo_id}`);
});

rotas.post("/laudos/:laudo_id{[0-9]+}/conferencia", async (c) => {
  await usuarioCom(c, "laudo.criar");
  const tx = c.get("tx");
  const laudo_id = _id(c);
  const f = await formulario(c);
  if (!f.tem("motivo")) throw new ErroHttp(422, "Campo obrigatório ausente: motivo");
  await tx
    .update(laudo_tecnico)
    .set({ data_ultima_conferencia: datas_br.hoje(), motivo_ultima_conferencia: f.texto("motivo") })
    .where(eq(laudo_tecnico.id, laudo_id));
  return redirecionar(c, `/laudos/${laudo_id}?mensagem=${encodeURIComponent("Conferência registrada.")}`);
});

rotas.post("/laudos/:laudo_id{[0-9]+}/superar", async (c) => {
  const usuario = await usuarioLogado(c);
  const tx = c.get("tx");
  const laudo_id = _id(c);
  const f = await formulario(c);
  if (!f.tem("motivo")) throw new ErroHttp(422, "Campo obrigatório ausente: motivo");
  const [laudo] = await tx.select().from(laudo_tecnico).where(eq(laudo_tecnico.id, laudo_id));
  if (!laudo) return redirecionar(c, "/laudos");
  const substituto_id = f.texto("substituto_id");
  const [substituto] = substituto_id
    ? await tx.select().from(laudo_tecnico).where(eq(laudo_tecnico.id, Number(substituto_id)))
    : [];
  const derivados = await servico_parecer.marcar_laudo_superado(tx, laudo, usuario, f.texto("motivo"), substituto ?? null);
  return redirecionar(
    c,
    `/laudos/${laudo_id}?mensagem=${encodeURIComponent(`Laudo superado; ${derivados.length} parecer(es) em reavaliação.`)}`,
  );
});
